File: internal/core/tags.go
package core

import (
	"os"
	"strings"

	"gopkg.in/yaml.v3"
)

type TagsFile struct {
	Categories []string `yaml:"categories"`
	Energy     []string `yaml:"energy"`
}

type TagStore struct {
	FilePath string
	Tags     TagsFile
}

// NewTagStore loads and parses the tags.yaml file at filePath.
func NewTagStore(filePath string) (*TagStore, error) {
	ts := &TagStore{FilePath: filePath}

	data, err := os.ReadFile(filePath)
	if err != nil {
		return nil, err
	}

	if err := yaml.Unmarshal(data, &ts.Tags); err != nil {
		return nil, err
	}

	return ts, nil
}

// AllTags returns all tags from all groups flattened into a single slice.
func (ts *TagStore) AllTags() []string {
	all := make([]string, 0, len(ts.Tags.Categories)+len(ts.Tags.Energy))
	all = append(all, ts.Tags.Categories...)
	all = append(all, ts.Tags.Energy...)
	return all
}

// IsValid checks if a tag exists (case-insensitive).
func (ts *TagStore) IsValid(tag string) bool {
	lower := strings.ToLower(tag)
	for _, t := range ts.AllTags() {
		if strings.ToLower(t) == lower {
			return true
		}
	}
	return false
}

// FuzzyMatch finds the closest matching tag using Levenshtein distance.
// Returns empty string if no match is within distance 3.
func (ts *TagStore) FuzzyMatch(input string) string {
	lower := strings.ToLower(input)
	bestMatch := ""
	bestDist := 4 // threshold: must be <= 3

	for _, tag := range ts.AllTags() {
		d := levenshtein(lower, strings.ToLower(tag))
		if d < bestDist {
			bestDist = d
			bestMatch = tag
		}
	}

	return bestMatch
}

// AddTag adds a new tag to the specified group and saves to disk.
// Group defaults to "categories" if empty.
func (ts *TagStore) AddTag(tag string, group string) error {
	tag = strings.ToLower(tag)
	if group == "" {
		group = "categories"
	}

	switch group {
	case "categories":
		ts.Tags.Categories = append(ts.Tags.Categories, tag)
	case "energy":
		ts.Tags.Energy = append(ts.Tags.Energy, tag)
	}

	return ts.Save()
}

// Save writes the current tags back to tags.yaml.
func (ts *TagStore) Save() error {
	data, err := yaml.Marshal(&ts.Tags)
	if err != nil {
		return err
	}
	return os.WriteFile(ts.FilePath, data, 0644)
}

// GroupedTags returns tags organized by group name.
func (ts *TagStore) GroupedTags() map[string][]string {
	return map[string][]string{
		"categories": ts.Tags.Categories,
		"energy":     ts.Tags.Energy,
	}
}

// levenshtein computes the Levenshtein distance between two strings.
func levenshtein(a, b string) int {
	la, lb := len(a), len(b)
	if la == 0 {
		return lb
	}
	if lb == 0 {
		return la
	}

	// Use a single row for space efficiency.
	prev := make([]int, lb+1)
	for j := 0; j <= lb; j++ {
		prev[j] = j
	}

	for i := 1; i <= la; i++ {
		curr := make([]int, lb+1)
		curr[0] = i
		for j := 1; j <= lb; j++ {
			cost := 1
			if a[i-1] == b[j-1] {
				cost = 0
			}
			curr[j] = min(
				prev[j]+1,      // deletion
				curr[j-1]+1,    // insertion
				prev[j-1]+cost, // substitution
			)
		}
		prev = curr
	}

	return prev[lb]
}

func min(vals ...int) int {
	m := vals[0]
	for _, v := range vals[1:] {
		if v < m {
			m = v
		}
	}
	return m
}

File: src/core/tags.ts
// src/core/tags.ts
//
// Tag validation and fuzzy matching using Levenshtein distance.

/**
 * Computes the Levenshtein edit distance between two strings.
 * Uses a single-row (two-row rolling) optimization for space efficiency.
 */
export function levenshtein(a: string, b: string): number {
  const la = a.length;
  const lb = b.length;

  if (la === 0) return lb;
  if (lb === 0) return la;

  let prev: number[] = [];
  for (let j = 0; j <= lb; j++) {
    prev[j] = j;
  }

  for (let i = 1; i <= la; i++) {
    const curr: number[] = new Array(lb + 1);
    curr[0] = i;
    for (let j = 1; j <= lb; j++) {
      const cost = a[i - 1] === b[j - 1] ? 0 : 1;
      curr[j] = Math.min(
        prev[j] + 1,      // deletion
        curr[j - 1] + 1,  // insertion
        prev[j - 1] + cost // substitution
      );
    }
    prev = curr;
  }

  return prev[lb];
}

/**
 * TagValidator validates tags against a known set and provides fuzzy matching.
 * All comparisons are case-insensitive.
 */
export class TagValidator {
  private knownTags: string[];

  constructor(knownTags: string[]) {
    this.knownTags = [...knownTags];
  }

  /**
   * Returns true if the tag is in the known set (case-insensitive).
   */
  isValid(tag: string): boolean {
    if (tag === '') return false;
    const lower = tag.toLowerCase();
    return this.knownTags.some((t) => t.toLowerCase() === lower);
  }

  /**
   * Returns the closest known tag within Levenshtein distance <= 3,
   * or null if no match is within the threshold.
   * Comparison is case-insensitive; the original casing of the known tag is returned.
   */
  fuzzyMatch(input: string): string | null {
    const lower = input.toLowerCase();
    let bestMatch: string | null = null;
    let bestDist = 4; // threshold: must be <= 3

    for (const tag of this.knownTags) {
      const d = levenshtein(lower, tag.toLowerCase());
      if (d < bestDist) {
        bestDist = d;
        bestMatch = tag;
      }
    }

    return bestMatch;
  }

  /**
   * Adds a tag to the known set. No-op if the tag already exists (case-insensitive).
   */
  addTag(tag: string): void {
    if (!this.isValid(tag)) {
      this.knownTags.push(tag);
    }
  }
}
